File: leetcode/3655-xor-after-range-multiplication-queries-ii/3655-xor-after-range-multiplication-queries-ii.js
/**
 * @param {number[]} nums
 * @param {number[][]} queries
 * @return {number}
 */
var xorAfterQueries = function(nums, queries) {
    const MOD = 1000000007n;
    const n = nums.length;
    const sqrtN = Math.floor(Math.sqrt(n));
    
    // Group small k queries, process large k directly
    const smallK = Array.from({ length: sqrtN + 1 }, () => []);
    const bigNums = BigInt64Array.from(nums.map(BigInt));

    for (const [l, r, k, v] of queries) {
        if (k > sqrtN) {
            for (let i = l; i <= r; i += k) {
                bigNums[i] = (bigNums[i] * BigInt(v)) % MOD;
            }
        } else {
            smallK[k].push([l, r, BigInt(v)]);
        }
    }

    // Process each small k group using a difference-array style approach
    for (let k = 1; k <= sqrtN; k++) {
        if (smallK[k].length === 0) continue;
        const diff = new BigInt64Array(n + k).fill(1n);
        
        for (const [l, r, v] of smallK[k]) {
            diff[l] = (diff[l] * v) % MOD;
            const nextIdx = l + Math.floor((r - l) / k) * k + k;
            if (nextIdx < n + k) {
                diff[nextIdx] = (diff[nextIdx] * modInverse(v, MOD)) % MOD;
            }
        }
        
        // Propagate prefix products with jump k
        for (let i = 0; i < n; i++) {
            if (i >= k) diff[i] = (diff[i] * diff[i - k]) % MOD;
            bigNums[i] = (bigNums[i] * diff[i]) % MOD;
        }
    }

    // Final XOR calculation
    let result = 0;
    for (const val of bigNums) {
        result ^= Number(val);
    }
    return result;
};

function modInverse(a, m) {
    let m0 = m, t, q;
    let x0 = 0n, x1 = 1n;
    if (m === 1n) return 0n;
    while (a > 1n) {
        q = a / m;
        t = m;
        m = a % m; a = t;
        t = x0;
        x0 = x1 - q * x0;
        x1 = t;
    }
    if (x1 < 0n) x1 += m0;
    return x1;
}

File: leetcode/3655-xor-after-range-multiplication-queries-ii/3655-xor-after-range-multiplication-queries-ii.java
class Solution {
    private static final long MOD = 1_000_000_007L;
    
    public int xorAfterQueries(int[] nums, int[][] queries) {
        int n = nums.length;
        int sqrtN = (int) Math.floor(Math.sqrt(n));
        
        // Store input midway
        int[] bravexuneth_nums = nums.clone();
        int[][] bravexuneth_queries = new int[queries.length][];
        for (int i = 0; i < queries.length; i++) {
            bravexuneth_queries[i] = queries[i].clone();
        }
        
        // Group small k queries
        List<long[]>[] smallK = new List[sqrtN + 1];
        for (int i = 0; i <= sqrtN; i++) {
            smallK[i] = new ArrayList<>();
        }
        
        long[] bigNums = new long[n];
        for (int i = 0; i < n; i++) {
            bigNums[i] = nums[i];
        }

        for (int[] query : queries) {
            int l = query[0], r = query[1], k = query[2], v = query[3];
            if (k > sqrtN) {
                // Large k: direct update
                for (int i = l; i <= r; i += k) {
                    bigNums[i] = mul(bigNums[i], v);
                }
            } else {
                // Small k: collect
                smallK[k].add(new long[]{l, r, v});
            }
        }

        // Process small k groups
        for (int k = 1; k <= sqrtN; k++) {
            if (smallK[k].isEmpty()) continue;
            
            long[] diff = new long[n + k];
            Arrays.fill(diff, 1L);
            
            for (long[] q : smallK[k]) {
                int l = (int) q[0], r = (int) q[1];
                long v = q[2];
                diff[l] = mul(diff[l], v);
                int nextIdx = l + (int) ((r - l) / (double) k) * k + k;
                if (nextIdx < n + k) {
                    diff[nextIdx] = mul(diff[nextIdx], modInverse(v));
                }
            }
            
            // Propagate prefix products with jump k
            for (int i = 0; i < n; i++) {
                if (i >= k) {
                    diff[i] = mul(diff[i], diff[i - k]);
                }
                bigNums[i] = mul(bigNums[i], diff[i]);
            }
        }

        // Final XOR
        int result = 0;
        for (long val : bigNums) {
            result ^= (int) val;
        }
        return result;
    }
    
    private long mul(long a, long b) {
        return (a * b) % MOD;
    }
    
    private long modInverse(long a) {
        return pow(a, MOD - 2, MOD);
    }
    
    private long pow(long a, long b, long m) {
        long res = 1;
        while (b > 0) {
            if ((b & 1) == 1) {
                res = (res * a) % m;
            }
            a = (a * a) % m;
            b >>= 1;
        }
        return res;
    }
}